Fix TXT record generation missing quotes

woutergd opened the pull request:
When using this OctoDNS Provider within Cloudalre, the TXT-records are generated without quotes around the content. This will generate a warning in Cloudflare, indicating that those quotes are required. However, due to the way OctoDNS works, it is impossible to add these quotes in the Yaml Configuration.

This change will add the quotes in the preparation function for TXT records, so that the quotes will always be added and resolving the warnings in the Cloudflare Dashboard.

The exact warning message cloudflare gives:
`The content field of TXT records must be in quotation marks. Cloudflare may add quotation marks on your behalf, which will not affect how the record works.`

It links to the following informational page: https://www.cloudflare.com/learning/dns/dns-records/dns-txt-record/, which also states that TXT records need to be in quotes: `However, TXT records still store their content within quotation marks ("")`

This fixes also the issue that is reported in #115

ross commented on octodns_cloudflare/__init__.py:
```diff
@@ -284,7 +284,7 @@ def _data_for_TXT(self, _type, records):
             'type': _type,
             'values': [
-                r.get('content', '').replace(';', '\\;') for r in records
+                f"\"{r.get('content', '').replace(';', '\\;')}\"" for r in records
```
This is getting complicated enough to read that it probably makes sense to break it up into a couple steps:

Maybe something like:

```python
values = [r.get('content', '').replace(';', '\\;') for r in records]
values = [f'"{v}"' for v in values]
...
    'values': values,
```

woutergd replied:
@ross thanks for the feedback, modified it. 

I switched back to double quotes as enclosure for these strings, as the Cloudflare API throws errors when content is sent with strings in single quotes in the JSON (I think it is even invalid json at that point). To avoid any confusion if the backslash is added or not as escape character, I think this is more clear than switching back to single quotes.

ross commented on .github/workflows/main.yml:
```diff
@@ -1,5 +1,5 @@
 name: octoDNS CloudflareProvider
-on: [pull_request]
+on: [push]
```
This shouldn't change.

ross reviewed (approved):
Looks good.

woutergd commented:
@ross  was running some tests in my own fork, apparently it caused some notifications for you I guess, wasn't what I intended.

I have updated the PR, now all tests are passing. Moved the main code back into single quotes, but found out that while the quotes are passed into the records, these are not outputted anymore afterwards when generating the zones. To work around this, I added a change in the file at the location where the records are prepared for pushing to the server to also add those quotes.

Just let me know if any changes are wanted!

ross commented:
I didn't step back and think about what you were saying enough. octoDNS's YAML config should not contain double quotes on values. That's really a PITA to get to happen with YAML, as you found, and would make it nearly impossible for non super technical users to work with the YAML.

The `_data_for_*` methods all translate the information coming from CF into the form that octoDNS requires and should not have containing quotes on TXT values.

If the CF API/UI calls are complaining about the lack of enclosing `"` that would need addressed when the octoDNS values are being converted into the form sent to CF's API. That all happens in the `_contents_for_*` functions, e.g.

https://github.com/octodns/octodns-cloudflare/blob/6edf23f8b8612a91c7a116905a413a04b718be14/octodns_cloudflare/__init__.py#L762-L764

That'd be the location where quoting needs to happen before it's sent up to their API.  

Where are you seeing the warnings? That should help me double check my understanding of what's up.

woutergd commented:
@ross the warnings show up in the DNS configuration in Cloudflare Dashboard (so the main UI). See also the screenshots in #115. The message states that TXT records should always be encapsulated within "" (see the exact message at the start of this PR). 

As cross-check, I also did a check on RFC specification, that defines the content of this field to be defines in the following way (page 35 of RFC1035):

<character-string> is expressed in one or two ways: as a contiguous set
of characters without interior spaces, or as a string beginning with a "
and ending with a ".  Inside a " delimited string any character can
occur, except for a " itself, which must be quoted using \ (back slash).

So it seems that CF is very strict by always requiring the quotes, but at least it should be possible to pass them (or automatically pass them, which I proposed in this PR). I understand that manually adding those quotes in yaml can be cumbersome and difficult to manage, but it should be relatively easy to do this in the processing I expect? In the end we should be able to add them in one way or another to comply with the specification. Probably every provider will implement this different and maybe add them by themselves, so adding them in the main repo for all providers would probably break stuff, but at least CF experts it to be passed in the API calls.

ross commented:
OK. I don't think CF complained in the past, but it sounds like it probably always expected the " and the provider wasn't doing it.  As mentioned in the previous comment that would be addressed in the `_contents_for_*` functions rather than `_data_for_*`. Everything should be coming out of CF into the form that octoDNS expects already. 

octoDNS's schema ignores the quotes, as they'd be a PITA to deal with in YAML, and it also avoids forcing the user to deal with the details of value length by dealing with the chunking. 

Adding the leading and trailing `"` should be easy in `_contents_for_*`. If CF wants things chunked dealing with that may be a bit more involved and will definitely require a bunch of testing and debugging. Feel free to limit this PR to the `"` part you set out to deal with and the other half can be left for later investigation.

Sorry I didn't grok the details earlier.

woutergd commented:
@ross  thanks for your update. I modified it to use the contents_for, apparently I missed that one when I started proposing this change. I kept the modified return files in the tests, as this is also how I retrieve the contents from CF at the moment when I perform a GET-request on TXT-records that has those quotes.

ross commented:
Looks good. 

You can run everything locally with 

```console
$ ./script/bootstramp
```

To get all the deps set up and 

```console
$ ./script/cibuild
```

to run to run the lints and tests. The coverage output will be stored in `htmlcov` after that. 

I'm tied up for the rest of today. Will take a look tomorrow morning and figure out what's up if you haven't gotten the chance to before me.

woutergd commented:
@ross Thanks for pointing out those scripts, didn't noticed that I could use those to test it completely. I updated the unit tests, coverage is now 100% again. 

I also added a .gitattributes file for the script-files, had some issues with line endings as I cloned the repo on Windows instead of Linux. With WSL it didn't run those scripts correctly, as it moved into CRLF line endings instead of LF what bash needs. With this file, every next clone will be preserving the correct line endings for those script-files, so they will run without issues.

If any other changes are necessary, just let me know.

ross commented on .gitattributes:
```diff
@@ -0,0 +1 @@
+script/* text eol=lf
```
I think this would fall into a similar category as the note in the main repo's `.gitignore` file. https://github.com/octodns/octodns/blob/100c2959036c5ffcecbbc8ed7d21b745d086f38e/.gitignore#L1-L4.

Since it's OS/editor specific it should be something that gets added to the user globals, in this case in `$HOME/.config/git/attributes`. 

If all the different editor and OS tweaks get added to the repo(s) things get big and complicated quickly and those things would need to be copied across all ~30 repos.

ross reviewed (commented):
Minor bit, changes/code and tests look good.

ross commented:
Hopefully you don't mind me pulling the `.gitattributes` file, reasoning is up in https://github.com/octodns/octodns-cloudflare/pull/134#discussion_r2020324325.

Thanks for the find & PR. Sorry again that the route was a bit more circuitous than it could have been.
